### ISSUE TITLE
1.0.8 deadlock?

### ISSUE THREAD

**Issue by nerdralph** [Genoil/cpp-ethereum#35]
Built 1.0.8 on Ubuntu 14.04.  ethminer hangs after the seedhash line.  0% cpu use, so I suspect a deadlock in the new lock/mutex code added.  1.0.7 works fine.
Using  -E old -G --cl-global-work 16384 -S
[OPENCL]:Creating one big buffer for the DAG
[OPENCL]:Loading single big chunk kernels
[OPENCL]:Mapping one big chunk.
[OPENCL]:Creating buffer for header.
[OPENCL]:Creating mining buffer 0
[OPENCL]:Creating mining buffer 1
[OPENCL]:Creating buffer for header.
[OPENCL]:Creating mining buffer 0
[OPENCL]:Creating mining buffer 1
  ℹ  04:18:17|gpuminer0  seedhash #95d53b63d0b53159

**Comment by Genoil** [comment]
And if you disable the work-time-out code? Could be that there's a difference between Windows and Linux. You only built on Linux I assume.

**Comment by nerdralph** [comment]
If you had a -DNO_TIMEOUT it would be easier to test without the timeout code (or even have --work-timeout 0 to disable it).  For the time being I've made my changes against my 1.0.7 fork and am still doing testing.
And yes, I only built on Linux.

**Comment by Genoil** [comment]
Thanks, I'll see if I can replicate the behaviour.

**Comment by Genoil** [comment]
Running 1.0.8 on an Ubuntu 14.04 Server Amazon AWS G2 instance (NVidia GRID K520) now in stratum mode on ethermine. No deadlock here. Both OpenCL and CUDA work fine.

**Comment by nerdralph** [comment]
I really don't like mutexes since they are *very* tricky to debug.  Wherever possible I prefer serialization.
I cloned a clean 108 branch, and built on 14.04.  Tried running it on a 4-card rig, and got the deadlock after the first "gpuminer0  seedhash #95d53b63d0b53159".  I used --opencl-devices to run on just 2 cards and added -v 9.  It worked fine.  Tried again with all 4 cards and -v 9 and it worked.  Removed -v 9 and tried again, and it works.

So I still suspect a deadlock, but I'll have to dig through the code before I can provide a theory.
